plan.py

```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
+import multiprocessing as mp
+from dataclasses import dataclass, field, asdict, is_dataclass
+from typing import Any, Callable, Dict, List, Literal, Optional, Protocol, Set, Tuple, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from promise import ScopedConstraint
+
+Status = Literal["pending", "running", "success", "failed", "terminated"]
+
+@dataclass(slots=True)
+class TaskInstance:
+    id: str #UUID
+    spec_name: str
+    inputs: Dict[str, Any]
+    status: Status = "pending"
+    parent_id: Optional[str] = None
+    labels: Set[str] = field(default_factory=set)
+    started_at: Optional[float] = None
+    ended_at: Optional[float] = None
+    outputs: Any = None
+
+Edge = Tuple[str, str]  # (from_id, to_id)
+
+@dataclass(slots=True)
+class Plan:
+    nodes: Dict[str, TaskInstance] = field(default_factory=dict) #maps node_id to TaskIntance's
+    edges: Set[Edge] = field(default_factory=set)
+    promises: List["ScopedConstraint"] = field(default_factory=list)
+
+    #returns a list of node_id
+    def predecessors(self, node_id: str) -> List[str]:
```

Comment:
```suggestion
    def predecessors(self, node_id: str) -> list[str]:
```

also, this can be precomputed and updated each time you add an edge (o.w. its O(|edges|) for each call)